### PR TITLE
feat(data): add 3D motion spatial recall dataset

### PR DESCRIPTION
**Pull request by Dafidofff**
## Summary

Add a 3D motion spatial-recall task: a resized image translates and optionally makes quarter turns through a cubic video block, with depth representing time. The model receives that block on a larger canvas and must reproduce it at a reserved back-bottom-right readout corner. For example, a 4×4 digit becomes a 6×6×6 target in a 32×32×32 canvas.

Includes `SpatialRecall3DMotionDataset`, its Lightning datamodule with volume/sequence batch layouts, usage examples, and 148 CPU tests. Fixed and random placements reserve the readout. The wrapper derives its input and target channel counts from the base image channels, including before setup, and supports grayscale and RGB sources.

Motion and loader generators use seeds derived from the base seed, global rank and stream ID; ranks receive distinct motion streams with or without loader workers, while base dataset splitting keeps its original seed. Rounded sweep increments obey `ceil(max_step)`, including the two-slice half-to-even rounding case. These fixes change generated motion for some historical seeds. The README distinguishes volume cropping from the explicit cube-index gather required for sequence readout.

This contribution contains the motion copy dataset, documentation and tests. It has no model changes, training results or colour-conditioning variant. Existing spatial-recall APIs are unchanged.

## Environment setup

Validated using an existing Python 3.12 environment with PyTorch 2.9.0+cu128 and pytorch-lightning. No dependency installation or dataset download was needed. CPU checks do not require a CUDA driver.

## Pre-commit

- All hooks passed on the three changed files, including Ruff, formatting, secret detection and license checks.

## Test plan

- [x] `python -m pytest tests/test_spatial_recall_motion.py -q -o addopts=''` — 148 passed.
- [x] Fixed/random source placement, reserved readout, exact copy labels, seeded sequences, blank images, invalid arguments, split loaders and both batch layouts.
- [x] Sweep bounds across lengths 1, 2, 3, 6 and 16, eight step limits, three motion ranges and ten seeds; translation actually changes frames.
- [x] Distinct and repeatable motion streams for simulated ranks 0 and 1, across all splits, with zero workers and real two-worker loaders.
- [x] One-step, two-process Lightning CPU DDP smoke run completed and verified distinct motion seeds on both ranks.
- [x] RGB channel metadata before setup and matching projection/batch dimensions; sequence gather matches the volume crop exactly.
- [x] Standalone validation exercises EMNIST's real stage dispatch with synthetic source data.
- [ ] Full GPU suite not run for this revision. The earlier attempt stopped during collection in existing `tests/modules/test_ckconv_nd_subq.py` because importing the CUDA extension raised `pynvml.NVMLError_DriverNotLoaded`.

## Documentation checklist

- [x] Existing module description and public class/method documentation updated.
- [x] Motion, clipping, rounding, rank/worker seeding and sequence readout contracts documented.
- [x] No new module under `nvsubquadratic/` or `experiments/`; no new docs-tracker row required.

### CODE REVIEW

File: experiments/datamodules/spatial_recall_dataset.py [hunk 1]
```diff
@@ -1944,6 +1945,561 @@ def on_before_batch_transfer(self, batch, dataloader_idx) -> dict:
...
+        return DataLoader(
+            dataset,
+            batch_size=self.batch_size,
+            shuffle=shuffle,
+            num_workers=self.num_workers,
+            pin_memory=self.pin_memory,
+            drop_last=drop_last,
+            generator=self._generator,
+            persistent_workers=self.num_workers > 0,
+            worker_init_fn=_seed_motion_worker,
```
Comment (farhadrgh) [inline experiments/datamodules/spatial_recall_dataset.py:2439]: **Every DDP rank draws the identical motion stream.**

Setting `worker_init_fn=_seed_motion_worker` suppresses Lightning's rank-aware seeding: `lightning_fabric.utilities.data._auto_add_worker_init_fn` only installs `pl_worker_init_function` when `dataloader.worker_init_fn is None`, so with `seed_everything(config.seed, workers=True)` (which `experiments/run.py` does call) the rank-aware path is skipped here.

The motion RNG chain is then rank-identical: `_generator = torch.Generator().manual_seed(self._seed)` (line 2326) is the same on every rank, so DataLoader's `_base_seed` — and therefore each `worker.seed` — matches across ranks.

On 8 GPUs, rank 0 worker 0 and rank 7 worker 0 produce the same sequence of `(q0, n_turns, direction, sweep gamma/flip)` draws, so the k-th sample on every rank gets the same motion path applied to a different digit. Augmentation diversity effectively drops to 1/world_size, and the README's "Loader workers receive distinct PyTorch worker seeds" holds only *within* a rank.

`experiments/datamodules/mnist.py:196` documents the repo convention of relying on `seed_everything(workers=True)` rather than a custom `worker_init_fn`, for this reason.

Fix: call `pl_worker_init_function(worker_id, rank)` inside `_seed_motion_worker` and seed the dataset generator from `torch.initial_seed()`, or mix `trainer.global_rank` into `self._generator`'s seed.

**Review by farhadrgh — COMMENTED** [review]
Review pass on the motion dataset — mostly focused on correctness of the sampling invariants and the DDP path. Four findings below that I think are worth addressing before merge; a longer list of quality/perf items (duplicated datamodule boilerplate, the precomputed position table, redundant per-slice resizes, a missing "does it actually move" test) I can post separately if useful.

Nice work on the test suite — 34 tests all pass locally, and the README example runs as written.

File: experiments/datamodules/spatial_recall_dataset.py [hunk 2]
```diff
@@ -1944,6 +1945,561 @@ def on_before_batch_transfer(self, batch, dataloader_idx) -> dict:
...
+    x = limit * ease
+    if float(torch.rand(1, generator=generator)) < 0.5:
+        x = limit - x  # random sweep direction
+    if length > 1:
+        step = float((x[1:] - x[:-1]).abs().max())
+        if step > max_step:
+            # Shrink wide sweeps about their centre before voxel rounding.
+            centre = limit / 2.0
+            x = centre + (x - centre) * (max_step / step)
+    return torch.clamp(torch.round(x), 0, limit).long()
```
Comment (farhadrgh) [inline experiments/datamodules/spatial_recall_dataset.py:1985]: **The sweep can emit a step of `2 * max_step`, double the documented cap.**

```python
_random_sweep_offsets(length=2, limit=2, max_step=1, generator=torch.Generator().manual_seed(0))
# -> [0, 2]    # step of 2, but ceil(max_step) == 1
```

The shrink at lines 1979-1984 scales the path so its largest *continuous* step equals `max_step` exactly; `torch.round` (round-half-to-even) then maps 0.5 -> 0 and 1.5 -> 2, pushing the endpoints one more voxel apart each.

I swept `length` in 2..16 x `limit` 0..31 x `max_step` in {0, 0.25, 0.5, 1, 1.5, 2, 2.5, 3} x 200 seeds: **every** seed at `length=2` with an integer `max_step` violates the bound (`max_step=1` -> step 2, `max_step=3` -> step 4). So `block_size=2` silently doubles the advertised speed cap.

`tests/test_spatial_recall_motion.py::test_sweeps_are_bounded_monotone_and_obey_rounded_step_cap` misses this because it hardcodes `length=6`.

Fix: shrink to `max_step` minus a rounding margin, or clamp the diffs of the rounded sequence.

File: examples/spatial_recall_v2/motion_3d/README.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    canvas_size=32,
+    data_type="volume",
+)
+# Explicitly call module.prepare_data() to download EMNIST if needed,
+# then module.setup("fit") to construct the train/validation datasets.
+```
+
+The batch-transfer hook returns `{"input": x, "label": y, "condition": None}`.
+With `data_type="volume"`, shapes are `[B, S, S, S, 1]` and `[B, b, b, b, 1]`;
+with `"sequence"`, they are `[B, S**3, 1]` and `[B, b**3, 1]`. A regression
```
Comment (farhadrgh) [inline examples/spatial_recall_v2/motion_3d/README.md:82]: **This readout recipe only works for `data_type="volume"`.**

In `sequence` mode `on_before_batch_transfer` flattens as `b (d h w) c`, so token index = `d*S^2 + h*S + w`. The label is the `block_size^3` cube, but the last `b^3` tokens of the canvas are indices `>= S^3 - b^3` — the tail of the `d = S-1` plane, not `canvas[:, S-b:, S-b:, S-b:]`.

Concretely: `ResidualNetwork._get_readout_region` on a `[B, S^3, C]` input has `spatial_ndim == 1`, so it rejects `target_size=[b, b, b]` with a `ValueError`, while `target_size=(b**3,)` crops the wrong tokens entirely.

So a user following this README with `data_type="sequence"` — which the module offers, the defaults document, and the tests assert shapes for — trains against a label that doesn't correspond to the cropped region.

Either document that sequence mode needs a gather over the cube's flat indices, or drop the sequence option.

File: experiments/datamodules/spatial_recall_dataset.py [hunk 4]
```diff
@@ -1944,6 +1945,561 @@ def on_before_batch_transfer(self, batch, dataloader_idx) -> dict:
...
+        self._seed: Optional[int] = None
+
+        # Generators.
+        self._generator: Optional[torch.Generator] = None
+        self._train_generator: Optional[torch.Generator] = None
+        self._val_generator: Optional[torch.Generator] = None
+        self._test_generator: Optional[torch.Generator] = None
+
+        # Input/output channels.
+        self.input_channels = 1
```
Comment (farhadrgh) [inline experiments/datamodules/spatial_recall_dataset.py:2305]: **`input_channels`/`output_channels` are hardcoded to 1, but the dataset passes the base image's channel count through.**

`SpatialRecall3DMotionDataset.__getitem__` sets `num_channels = block.shape[0]` from the source image, so wrapping any 3-channel base datamodule yields batches of `[B, S, S, S, 3]` while the datamodule still advertises `input_channels = 1`.

A config wiring `net.in_channels = '${dataset.input_channels}'` then builds a `Linear(1, hidden_dim)` and the forward pass fails with a shape mismatch at the in-projection.

Fix: derive both from the base datamodule in `_extract_base_properties` (`base.input_channels`), or validate that the base is single-channel and raise at construction.

**Review by farhadrgh — APPROVED** [review]

**Review by saeepaliwal — APPROVED** [review]
